src/ucp/api/ucp.h

```diff
@@ -1738,6 +1788,36 @@ struct ucp_am_recv_param {
      * Endpoint, which can be used for the reply to this message.
      */
     ucp_ep_h           reply_ep;
+
+    /**
+     * Length of the whole message in bytes. Relevant for multi-fragment eager
+     * messages handled by data handlers registered without
+     * @a UCP_AM_FLAG_WHOLE_MSG flag.
+     */
+    size_t             total_length;
+
+    /**
+     * Offset of the message fragment in bytes relative to the beginning of
+     * overall message. Layout of the multi-fragment message is depicted below:
+     *        Multi-fragment message
+     *  +--------+--------+--------+--------+
+     *  |frag 1  |frag 2  |  ...   |frag N  |
+     *  +--------+--------+--------+--------+
+     *           |                 v
+     *           |               offset of the N-th fragment
+     *           v
+     *         offset of the 2-d fragment
+     */
+    size_t             frag_offset;
+
+
+    /**
+     * User data associated with multi-fragment message. Relevant for middle and
+     * last chunks of the multi-fragment message. This field contains the value
+     * passed to @ref ucp_am_recv_data_nbx invoked for the first fragment of the
+     * message, so that the rest fragments of the same message can be matched.
+     */
+    void               *user_data;
```

Comment:
maybe rename it to message_context ?

Reply:
last time we agreed that `user_data` is more representative than `message_context`